### PR TITLE
Various updates to `PregelNode` and `Pregel` classes. Update `validate.ts` to match Python implementation.

### PR DESCRIPTION
**Pull request by andrewnguonly**
### Summary
This PR includes various updates to the `PregelNode` and `Pregel` classes. They're smaller, isolated changes with the goal of reducing the number lines changed in future PRs that will include more business logic changes. For example, some of these changes are non-functional (e.g. renaming variables).

Additionally, the `validate.ts` file is updated to match the Python implementation.

### Implementation
1. `Pregel` changes:
  a. Update the class interface to match the Python implementation. For example, adding `streamMode` and removing `hidden`.
  b. Rename fields. For example, `input` --> `inputChannels`.
  c. Create standalone `validate()` method. Call it in constructor.
2. Update implementation of `Channel.writeTo()`.
3. `PregelNode` changes:
  a. Implement `getWriters()` and `getNode()` functions.
  b. Change `channels` from `Record<string, string> | string` to support array, `string[]`. This change has propagating changes in other files.
4. Update `validate.ts` to match the Python implementation.
5. Removed reserved channels functionality (i.e. `reserved.ts`).

### Next Steps
1. Migrate `Pregel._prepareNextTasks()`.
3. Migrate `Pregel.stream()` and `Pregel.invoke()`. This will be the bulk of business logic changes.
4. Apply changes from this PR: https://github.com/langchain-ai/langgraph/pull/366
5. Apply changes from this PR: https://github.com/langchain-ai/langgraph/pull/367

### Notes
1. I'm not sure why the commit history includes a bunch of extra commits. I think it's because I branched off a branch and not `main`.

### CODE REVIEW

File: langgraph/src/pregel/read.ts [hunk 1]
```diff
@@ -119,7 +120,7 @@ export class PregelNode<
 > extends RunnableBinding<RunInput, RunOutput, RunnableConfig> {
   lc_graph_name = "PregelNode";
 
-  channels: Record<string, string> | string;
+  channels: Record<string, string> | string[];
```
Comment (andrewnguonly) [inline langgraph/src/pregel/read.ts:123]: This type changes requires updating the implementation of `_prepareNextTasks()` and a bunch of tests.

File: langgraph/src/pregel/read.ts [hunk 2]
```diff
@@ -159,6 +160,47 @@ export class PregelNode<
     this.kwargs = kwargs ?? this.kwargs;
   }
 
+  getWriters(): Array<Runnable> {
```
Comment (andrewnguonly) [inline langgraph/src/pregel/read.ts:163]: The updated Python implementation of `_prepareNextTasks()` requires `getWriters()` and `getNode()`.

File: langgraph/src/pregel/index.ts [hunk 3]
```diff
@@ -198,51 +206,83 @@ export class Pregel
   // Because Pregel extends `Runnable`.
   lc_namespace = ["langgraph", "pregel"];
 
+  nodes: Record<string, PregelNode>;
+
   channels: Record<string, BaseChannel> = {};
 
-  output: string | Array<string> = "output";
```
Comment (andrewnguonly) [inline langgraph/src/pregel/index.ts:203]: `Pregel` interface updates:
1. `output` --> `outputChannels`
2. `input` --> `inputChannels`
3. `hidden` is removed
4. `interrupt` --> `interruptAfterNodes` and `interruptBeforeNodes`
5. `streamMode` is added (unused right now)
6. `streamChannels` is added
7. `autoValidate` is added (unused right now)

File: langgraph/src/pregel/index.ts [hunk 4]
```diff
@@ -360,13 +398,6 @@ export class Pregel
         checkpoint = await createCheckpoint(checkpoint, channels);
         await this.checkpointer.put(config, checkpoint);
       }
-
-      // interrupt if any channel written to is in interrupt list
-      if (
-        pendingWrites.some(([chan]) => this.interrupt?.some((i) => i === chan))
```
Comment (andrewnguonly) [inline langgraph/src/pregel/index.ts:366]: If this logic needs to be migrated, I haven't done so yet.

File: langgraph/src/pregel/index.ts [hunk 5]
```diff
@@ -84,13 +76,13 @@ export class Channel {
       );
     }
 
-    let channelMappingOrString: string | Record<string, string>;
+    let channelMappingOrString: string[] | Record<string, string>;
```
Comment (andrewnguonly) [inline langgraph/src/pregel/index.ts:79]: Rename `channelMappingOrString` to `channelMappingOrArray`.

File: langgraph/src/pregel/index.ts (hunk 4 above)
Comment (nfcampos) [reply langgraph/src/pregel/index.ts:366]: it does

Comment (nfcampos) [reply langgraph/src/pregel/index.ts:366]: but can be in next pr

**Review by nfcampos — APPROVED** [review]

**Review by nfcampos — APPROVED** [review]

**Pull request merged**